File: scripts/SCANB_modalities.R
#!/usr/bin/env Rscript
# Script: overview data modalities SCAN-B
# Author: Lennart Hohmann
# Date: 11.03.2025
#-------------------
# empty environment
rm(list=ls())
# set working directory to the project directory
setwd("~/PhD_Workspace/PredictRecurrence/")
#-------------------
# packages
#source("./scripts/src/")
if (!require("pacman")) install.packages("pacman")
pacman::p_load(ggplot2, ggVennDiagram, gridExtra)
#-------------------
# set/create output directories
output_path <- "./output/"
dir.create(output_path)
#-------------------
# input paths
infile_1 <- "./data/standardized/SCANB_sample_modalities.csv"
infile_2 <- "./data/standardized/SCANB_clinical.csv"
infile_3 <- "./data/standardized/SCANB_RNAseq_expression.csv"
infile_4 <- "./data/standardized/SCANB_RNAseq_mutations.csv"
infile_5 <- "./data/standardized/SCANB_DNAmethylation.csv"
#-------------------
# which clin group to run for
clin_group <- "ER+HER2-"
#-------------------
# output paths
outfile_1 <- paste0(output_path,"SCANB_Modalities_",clin_group,".pdf")
outfile_2 <- paste0(output.path,"SCANB_Modalities_",clin_group,".txt") 
#-------------------
# storing objects 
plot.list <- list() # object to store plots
txt.out <- c() # object to store text output, ggf. use capture.output()

#######################################################################
# load data
#######################################################################

sample_modalities <- read.csv(infile_1)
clinical <- read.csv(infile_2)
RNAseq_expr <- read.csv(infile_3) # nolint
RNAseq_mut <- read.csv(infile_4)
DNAmethyl <- read.csv(infile_5) # until the real data is loaded
names(DNAmethyl) <- c("Sample")

# Subgroup data
if (clin_group == "All") {
  sub_sample_modalities <- sample_modalities
  sub_clinical <- clinical
  sub_RNAseq_expr <- RNAseq_expr
  sub_RNAseq_mut <- RNAseq_mut
  sub_DNAmethyl <- DNAmethyl
  } else {
    sub_clinical <- subset(clinical, Group == clin_group)
    sub_sample_modalities <- subset(sample_modalities, Sample %in% sub_clinical$Sample)
    sub_RNAseq_expr <- RNAseq_expr[, c("Gene", intersect(colnames(RNAseq_expr), sub_clinical$Sample))]
    sub_RNAseq_mut <- RNAseq_mut[, c("Gene", intersect(colnames(RNAseq_mut), sub_clinical$Sample))]
    sub_DNAmethyl <- DNAmethyl[DNAmethyl$Sample %in% sub_clinical$Sample, ]
}

#######################################################################
# Pie chart FU cohort ClinGroups
#######################################################################

group_counts <- table(clinical$Group)[c("ER+HER2-","ER+HER2+","ER-HER2+","TNBC","Other")]
group_counts <- as.data.frame(group_counts)
group_counts$Percentage <- round((group_counts$Freq / sum(group_counts$Freq)) * 100,0)
# Plot the pie chart using ggplot2
plot.1 <- ggplot(data = group_counts, aes(x = "", y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", width = 1, show.legend = TRUE) + 
  coord_polar(theta = "y") +  # This makes the chart circular
  labs(title = "ER & HER2 status") +
  theme_void() +  # Remove the background grid
  theme(legend.position = "none",
    axis.text.x = element_blank(),  # Remove x-axis text
        axis.ticks = element_blank(),  # Remove axis ticks
        legend.title = element_blank()) +  # Remove legend title
  scale_fill_manual(values = c(
    "ER+HER2-" = "#abd9e9", 
    "ER+HER2+" = "#f1b6da", 
    "ER-HER2+" = "#d01c8b", 
    "TNBC" = "#d7191c",
    "Other" ="#bababa"
  )) +
  geom_text(aes(label = paste(Var1, "\n", Freq, " (", Percentage, "%)", sep = "")), 
            position = position_stack(vjust = 0.5), size = 4)

#######################################################################
# Whole FU cohort modalities
#######################################################################

# Prepare list of sample groups
venn_list <- list(
  "DNAmethyl" = modalities.df$Sample[modalities.df$DNAmethyl == 1],
  "RNAseq.mut" = modalities.df$Sample[modalities.df$RNAseq.mut == 1],
  "RNAseq.gex" = modalities.df$Sample[modalities.df$RNAseq.gex == 1]
)

# Create the Venn diagram
plot.1 <- ggVennDiagram(venn_list, label_alpha = 0) + 
  scale_fill_gradientn(colors = c("#F8BBD0", "#EC407A", "#C2185B"), 
                       values = c(0, 0.2, 1)) +
  theme_void() +  
  theme(legend.position = "none") +
  labs(
    title = paste0("SCAN-B; All; n=",length(anno$Sample)),
    subtitle = "With variants from RNAseq"
  )

# Prepare list of sample groups
# venn_list <- list(
#   "DNAmethyl" = modalities.df$Sample[modalities.df$DNAmethyl == 1],
#   "WGS.mut" = modalities.df$Sample[modalities.df$WGS.mut == 1],
#   "RNAseq.gex" = modalities.df$Sample[modalities.df$RNAseq.gex == 1]
# )

# # Create the Venn diagram
# plot.2 <- ggVennDiagram(venn_list, label_alpha = 0) + 
#   scale_fill_gradientn(colors = c("#F8BBD0", "#EC407A", "#C2185B"), 
#                        values = c(0, 0.2, 1)) + 
#   theme_void() +  
#   theme(legend.position = "none") +
#   labs(
#     title = paste0("SCAN-B; All; n=",length(anno$Sample),
#     subtitle = "With variants from WGS"
#   ))


#######################################################################
# ERpHER2n FU cohort modalities
#######################################################################

clin.group <- "ER+HER2-"

# Prepare list of sample groups
venn_list <- list(
  "DNAmethyl" = modalities.df$Sample[modalities.df$DNAmethyl == 1 & 
                                       modalities.df$Sample %in% anno$Sample[anno$Group==clin.group]],
  "RNAseq.mut" = modalities.df$Sample[modalities.df$RNAseq.mut == 1 & 
                                        modalities.df$Sample %in% anno$Sample[anno$Group==clin.group]],
  "RNAseq.gex" = modalities.df$Sample[modalities.df$RNAseq.gex == 1 & 
                                        modalities.df$Sample %in% anno$Sample[anno$Group==clin.group]]
)

# Create the Venn diagram
plot.3 <- ggVennDiagram(venn_list, label_alpha = 0) + 
  scale_fill_gradientn(colors = c("#F8BBD0", "#EC407A", "#C2185B"), 
                       values = c(0, 0.2, 1)) +
  theme_void() +  
  theme(legend.position = "none") +
  labs(
    title = paste0("SCAN-B; ER+HER2-; n=",
                   length(anno$Sample[anno$Group==clin.group])),
    subtitle = "With variants from RNAseq"
  )

# # Prepare list of sample groups
# venn_list <- list(
#   "DNAmethyl" = modalities.df$Sample[modalities.df$DNAmethyl == 1 & 
#                                        modalities.df$Sample %in% anno$Sample[anno$Group==clin.group]],
#   "WGS.mut" = modalities.df$Sample[modalities.df$WGS.mut == 1 & 
#                                      modalities.df$Sample %in% anno$Sample[anno$Group==clin.group]],
#   "RNAseq.gex" = modalities.df$Sample[modalities.df$RNAseq.gex == 1 & 
#                                         modalities.df$Sample %in% anno$Sample[anno$Group==clin.group]]
# )

# # Create the Venn diagram
# plot.4 <- ggVennDiagram(venn_list, label_alpha = 0) + 
#   scale_fill_gradientn(colors = c("#F8BBD0", "#EC407A", "#C2185B"), 
#                        values = c(0, 0.2, 1)) + 
#   theme_void() +  
#   theme(legend.position = "none") +
#   labs(
#     title = paste0("SCAN-B; ER+HER2-; n=",
#                    length(anno$Sample[anno$Group==clin.group])),
#     subtitle = "With variants from WGS"
#   )



#######################################################################
# venn ERpHER2n FU cohort treatment #TreatGroup
#######################################################################

venn_list <- list(
  "Endocrine" = anno$Sample[anno$Endo == 1 & anno$Group==clin.group],
  "Chemo" = anno$Sample[anno$Chemo == 1 & anno$Group==clin.group],
  "Immu" = anno$Sample[anno$Immu == 1 & anno$Group==clin.group])

# Create the Venn diagram
plot.6 <- ggVennDiagram(venn_list, label_alpha = 0) + 
  scale_fill_gradientn(colors = c("#7fbf7b", "#c2a5cf","#fdc086")) + 
  theme_void() +  
  theme(legend.position = "none") +
  labs(
    title = paste0("Treatment in ERpHER2n; n=",
                   length(anno$Sample[anno$Group==clin.group])),
    subtitle = paste0("None n=", 
    length(anno$Sample[anno$TreatGroup == "None" & anno$Group==clin.group]),
    "; Missing n=",length(anno$Sample[anno$TreatGroup == "Missing" & anno$Group==clin.group]))
  )

#######################################################################
# save plots to pdf
#######################################################################

# Set up the PDF output
pdf(file = outfile.1, onefile = TRUE, width = 8.27, height = 11.69)

# Arrange the plots in a grid (2x2)
# grid.arrange(plot.1, plot.2, plot.3, plot.4, 
#              plot.5, plot.6, ncol = 2, nrow = 3)
grid.arrange(plot.5,plot.1, plot.3, plot.6, ncol = 2, nrow = 2)
# Close the PDF device to save the file
dev.off()

File: scripts/SCANB_EDA/SCANB_modalities.py
#!/usr/bin/env python3
################################################################################
# Script: Overview of the SCAN-B Follow-up cohort (2010-2018)
# Author: Lennart Hohmann
# Date: 11.03.2025
################################################################################

# import 
import os
import sys
import missingno as msno
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import sklearn
from venn import venn
from matplotlib.backends.backend_pdf import PdfPages
import src.utils as my
my.test_func()

# set wd
os.chdir(os.path.expanduser("~/PhD_Workspace/PredictRecurrence/"))

################################################################################
################################################################################

# input paths
infile_1 = "./data/standardized/SCANB_sample_modalities.csv"
infile_2 = "./data/standardized/SCANB_clinical.csv"
infile_3 = "./data/standardized/SCANB_RNAseq_expression.csv"
infile_4 = "./data/standardized/SCANB_RNAseq_mutations.csv"
infile_5 = "./data/standardized/SCANB_DNAmethylation.csv"

# set subgroup to run the analyses
#clin_group = "All" # "All" or "ER+HER2-"
clin_group = "ER+HER2-"

fig_size = (8.27/2, 11.69/2)

# output paths
outfile_1 = f"./output/SCANB_Modalities_Surv_{clin_group}.pdf"

# plot file
pdf = PdfPages(outfile_1)

################################################################################
# load data
################################################################################

# read in data
sample_modalities = pd.read_csv(infile_1)
clinical = pd.read_csv(infile_2)
RNAseq_expr = pd.read_csv(infile_3)
RNAseq_mut = pd.read_csv(infile_4)
DNAmethyl = pd.read_csv(infile_5) # until the real data is loaded
DNAmethyl.columns = ["Sample"] # until the real data is loaded

# subgroup data
if clin_group == "All":
    sub_sample_modalities = sample_modalities
    sub_clinical = clinical
    sub_RNAseq_expr = RNAseq_expr
    sub_RNAseq_mut = RNAseq_mut
    sub_DNAmethyl = DNAmethyl
else:
    sub_clinical = clinical[clinical['Group']==clin_group]
    sub_sample_modalities = sample_modalities[sample_modalities['Sample'].isin(sub_clinical['Sample'])]
    sub_RNAseq_expr = RNAseq_expr.loc[:,
                                      RNAseq_expr.columns.isin(['Gene'] + list(sub_clinical["Sample"]))]
    sub_RNAseq_mut = RNAseq_mut.loc[:,
                                      RNAseq_mut.columns.isin(['Gene'] + list(sub_clinical["Sample"]))]
    sub_DNAmethyl = DNAmethyl.isin(sub_clinical['Sample'])


################################################################################
# Piechart of ER and HER2 status in SCAN-B
################################################################################

group_order = ["ER+HER2-", "ER+HER2+", "ER-HER2+", "TNBC", "Other"]
group_counts = clinical['Group'].value_counts()[group_order]
# rename the count column as 'Count'
group_counts_df = group_counts.reset_index(name='Count')
# Add a new column 'Percentage' and calculate the percentage for each group
group_counts_df = group_counts_df.assign(
    Percentage=lambda x: (x['Count'] / x['Count'].sum()) * 100
)
group_counts_df = group_counts_df.round({'Percentage': 0})
# Rename the 'index' column to 'Group'
group_counts_df = group_counts_df.rename(columns={'index': 'Group'})
# Plotting the pie chart
plt.figure(figsize=fig_size)
plt.pie(group_counts_df['Count'], 
        labels=group_counts_df['Group'] + '\n' + group_counts_df['Count'].astype(str) + ' (' + group_counts_df['Percentage'].astype(str) + '%)', 
        autopct='%1.0f%%', 
        colors=['#abd9e9', '#f1b6da', '#d01c8b', '#d7191c', '#bababa'], 
        startangle=90, 
        wedgeprops={'edgecolor': 'black'})
plt.title('ER & HER2 status')
plt.axis('equal')  # Equal aspect ratio ensures the pie chart is circular
pdf.savefig() # save to pdf
#plt.show()
plt.close()

################################################################################
# Available data modalities
################################################################################

# Create a list of sample groups based on conditions in the DataFrame
venn_dict = {
    "DNAmethyl": set(sub_sample_modalities[sub_sample_modalities['DNAmethylation'] == 1]['Sample']),
    "RNAseq_mut": set(sub_sample_modalities[sub_sample_modalities['RNAseq_mutations'] == 1]['Sample']),
    "RNAseq_gex": set(sub_sample_modalities[sub_sample_modalities['RNAseq_expression'] == 1]['Sample'])
}
plt.figure(figsize=fig_size)
venn(venn_dict)
plt.title(f"SCAN-B; {clin_group}; n={len(sub_sample_modalities['Sample'])}")
pdf.savefig() # save to pdf
#plt.show()
plt.close()

################################################################################
# In-depth look: Clinicopathological variables
################################################################################

summary = sub_clinical.describe(include='all')  # Summary of all columns (numerical and categorical)
#print(summary)

# Missing data
plt.figure()#figsize=fig_size[::-1])
msno.matrix(sub_clinical)
plt.title('Nullity Matrix')
plt.savefig(f"./output/SCANB_NullityMatrix_{clin_group}.png", pad_inches=0.1)
#pdf.savefig(pad_inches=0.1)  # Save to PDF
plt.close()

plt.figure()#figsize=fig_size[::-1])
msno.bar(sub_clinical)
plt.title('Nullity by Column')
pdf.savefig(pad_inches=0.1)  # Save to PDF
plt.close()

# number of events per outcome measure
# OS vs RFI
cross_tab = pd.crosstab(sub_clinical['OS_event'], sub_clinical['RFi_event'])
#print(cross_tab)
formatted_string = "; ".join([
    f"OS_event 0 n= {cross_tab.iloc[0,0]}",
    f"OS_event 1 n= {cross_tab.iloc[1,0]}",
    f"RFi_event 0 n= {cross_tab.iloc[0,1]}",
    f"RFi_event 1 n= {cross_tab.iloc[1,1]}"
])

#print(f"OS vs. RFI events: {formatted_string}.")
# Count the events
event_counts = sub_clinical[["OS_event", "RFi_event","DRFi_event"]].apply(pd.Series.value_counts).T
# plot bars in stack manner
ax = event_counts.plot(kind='bar', stacked=True, figsize=fig_size, 
                       rot=0, xlabel='Class', ylabel='Count')
for c in ax.containers:

    # Optional: if the segment is small or 0, customize the labels
    labels = [v.get_height() if v.get_height() > 0 else '' for v in c]
    
    # remove the labels parameter if it's not needed for customized labels
    ax.bar_label(c, labels=list(map(int, labels)), label_type='center')

plt.title(f"OS vs. RFI events: {formatted_string}.",fontsize=8)
plt.tight_layout()
pdf.savefig()  # Save to PDF
plt.close()

# treatments
treatment_counts = sub_clinical[["Endo","Chemo"]]
# Create a list of sample groups based on conditions in the DataFrame
venn_dict = {
    "Endo": set(sub_clinical[sub_clinical['Endo'] == 1]['Sample']),
    "Chemo": set(sub_clinical[sub_clinical['Chemo'] == 1]['Sample']),
    "Immu": set(sub_clinical[sub_clinical['Immu'] == 1]['Sample'])
}

plt.figure(figsize=fig_size)
venn(venn_dict)
plt.title(f"SCAN-B; {clin_group}; n={len(sub_clinical['Sample'])}")
pdf.savefig() # save to pdf
#plt.show()
plt.close()

pdf.close()


################################################################################
# Survival analyses and Censoring distribution
################################################################################
